docs/rfc/012-record-transaction-model.md

```diff
@@ -0,0 +1,243 @@
+The `const` inputs are compiled into the zero-knowledge circuit,
+so they can be ignored for our purpose here,
+leaving the non-`const` inputs and the output for consideration.
+
+The execution of `main` can be described as a mathematical function
+```
+main : Record x Record x Inputs -> Record x Record x Output
+```
+where `x` is cartesian product,
+`Record` is the set of possible records,
+`Inputs` is the set of possible inputs to `main`, and
+`Output` is the set of possible outputs from `main`.
+(These sets can be thought as "types", but mathematically we talk about sets.)
+That is, this mathematical function
+takes three inputs (the two old records and the `main` inputs)
+and returns three outputs (the two new records and the `main` output).
+While `Record` is fixed, i.e. it is the same for all Leo programs,
+both `Inputs` and `Output` differ based on the Leo input and output types of `main`.
+
+In the Leo code, in `main` or in functions called by `main`,
+the values in `Inputs` are accessed via the `main` parameters,
+while the old records are accessed via the special `input` variable,
+which provides access to the two old records and their slots,
+including the payloads that contain application-specific data.
+The picture for new records and values in `Output` is less clear from the documentation:
+experimentation suggests that the new records are obtained
+by serializing the output value in `Output` (which, recall, may be a tuple).
+
+It is important to note that the values in `Inputs` do not come from the two old records.
+Rather, they are private inputs, supplied by the developer
+when they compile the Leo program and generate the zero-knowledge proof.
+Indeed, as mentioned above, the processing of the transaction in the blockchain
+does not execute the Leo code, and thus does not need to know the values in `Inputs`.
+Rather, the blockchain has to verify a zero-knowledge proof asserting that
+there exist values in `Input`, known to the creator of the transaction,
+such that the execution of the Leo program's `main`
+on those values and on the old records
+yields the new records;
+this is, roughly speaking, the assertion proved in zero-knowledge.
+
+### Proposed Leo Program Execution Model
+
+The current model described above seems adequate overall, but we need to:
+1. Clarify how Leo code reads old records and writes new records.
+2. Generalize from one entry point (i.e. the `main` function) to multiple entry points, in line with the smart contract paradigm.
+
+Generalizing from one `main` entry point to multiple ones is conceptually easy.
+It means that, instead of implicitly designating `main` as the only entry point,
+we need a mechanism to explicitly designate one or more Leo functions as entry points.
+
+A simple approach could be to use an annotation like `@entrypoint` to designate _entry point functions_:
+```
+@entrypoint
```

Comment:
This seems like a pretty clean way to do it syntactically, reducing the number of syntactic forms in favor of "attributes" / "annotations", in favor of allowing more semantic passes to deal with them.